flush-in-loop-behavior.markdown

```diff
@@ -0,0 +1,48 @@
+
+
+Take this example:
+
+```typescript
+const authors = await em.find(Author, { id: ["1", "2"] } );
+await Promise.all(
+  authors.map(async (a) => {
+    const p = await em.findOneOrFail(Publisher, { id: a.publisher.id });
+    a.firstName = a.firstName + p.name;
+    await em.flush();
+  }),
+);
+```
+
+The execution flow, in terms of SQL statements, looks something like:
+
+* `em.find` issues Q1 `SELECT * FROM authors WHERE id IN (1, 2);`
+* Q1 returns two authors, `.map` creates lambdas for both
+* Lambda 1 issues Q2 `SELECT * FROM publishers WHERE id = 3`, `await`s
+* Lambda 2 issues Q3 `SELECT * FROM publishers WHERE id = 4`, `await`s
+* Q2 returns, Lambda 1 modifies it's `a`, calls `em.flush`
+* Lambda 1's `flush` issues Q4 `UPDATE authors SET first_name = ...`
+* Q3 returns, Lambda 2 modifies it's `a`, calls `em.flush`
+* Lambda 2's `flush` issues Q4 `UPDATE authors SET first_name = ...`
+
+We've `N+1`'d.
```

Comment:
I.e. Joist's general ethos/goal of "we'll bulk update as many operations as possible" didn't happen, where if this was a loop with, say, 50 authors, we'd probably end up with 50 `UPDATE` calls instead of a single `UPDATE`.

Granted, some/most ORMs don't do bulk `UPDATE`s at all, but it's part of Joist's mental model, so I think it's surprising to user's when this doesn't happen. I.e. a causal read of the `.map` it _seems_ like that `em.flush()` should happen "at the same time" / be N+1 safe between both lambdas.